internal/pkg/skuba/deployments/ssh/kubernetes.go

```diff
@@ -56,24 +58,102 @@ func kubernetesUploadSecrets(errorHandling KubernetesUploadSecretsErrorBehavior)
 	}
 }
 
-func kubernetesInstallNodePattern(t *Target, data interface{}) error {
+func kubernetesParseInterfaceVersions(data interface{}) (string, string, error) {
 	kubernetesBaseOSConfiguration, ok := data.(deployments.KubernetesBaseOSConfiguration)
 	if !ok {
-		return errors.New("couldn't access kubernetes base OS configuration")
+		return "", "", errors.New("couldn't access kubernetes base OS configuration")
 	}
 
 	v, err := version.ParseSemantic(kubernetesBaseOSConfiguration.CurrentVersion)
 	if err != nil {
-		return err
+		return "", "", err
 	}
 	currentVersion := kubernetes.MajorMinorVersion(v)
-	patternName := fmt.Sprintf("patterns-caasp-Node-%s", currentVersion)
+	updatedVersion := ""
 
 	if kubernetesBaseOSConfiguration.UpdatedVersion != "" {
-		updatedVersion := kubernetes.MajorMinorVersion(version.MustParseSemantic(kubernetesBaseOSConfiguration.UpdatedVersion))
-		patternName = fmt.Sprintf("patterns-caasp-Node-%s-%s", currentVersion, updatedVersion)
+		updatedVersion = kubernetes.MajorMinorVersion(version.MustParseSemantic(kubernetesBaseOSConfiguration.UpdatedVersion))
+	}
+	return currentVersion, updatedVersion, nil
+}
+
+func kubernetesFreshInstallAllPkgs(t *Target, data interface{}) error {
+	current, _, err := kubernetesParseInterfaceVersions(data)
+	if err != nil {
+		return err
+	}
+	var pkgs []string
+
+	// Standard packages for a new cluster
+	pkgs = append(pkgs, "+caasp-release", "skuba-update", "supportutils-plugin-suse-caasp", "cri-tools")
+	// Version specific
+	pkgs = append(pkgs, fmt.Sprintf("+kubernetes-%s-kubeadm", current))
+	pkgs = append(pkgs, fmt.Sprintf("+kubernetes-%s-kubelet", current))
+	pkgs = append(pkgs, fmt.Sprintf("+kubernetes-%s-client", current))
+	pkgs = append(pkgs, fmt.Sprintf("+cri-o-%s*", current))
+
+	_, _, err = t.zypperInstall(pkgs...)
+	return err
+}
+
+func kubernetesUpgradeStageOne(t *Target, data interface{}) error {
+	// Stage1 upgrades only kubeadm.
+	// zypper install -- -kubernetes-old-kubeadm +kubernetes-new-kubeadm
+	currentV, nextV, err := kubernetesParseInterfaceVersions(data)
+	if err != nil {
+		return err
+	}
+	if nextV == "" {
+		return errors.New("Incorrect upgrade version")
 	}
-	_, _, err = t.ssh("zypper", "--userdata", "skuba", "--non-interactive", "install", patternName)
+
+	var pkgs []string
+
+	if currentV == "1.17" {
+		// 1.17 is the last version included in CaaSP4. It's the tipping
+		// point where we changed our packaging.
+		// On 1.17 we can't remove kubernetes-1.17-kubeadm, because it doesn't exist.
+		// Removing kubeadm keeps kubelet alive.
+		// The rest needs to be removed on the next stage.
+		pkgs = append(pkgs, "-patterns-caasp-Node-1.17", "-kubernetes-kubeadm", "-cri-o-kubeadm-criconfig")
```

Comment:
Frankly, this makes me cringe. Removal of v4 patterns/packages should be handled by `Obsoletes` imho.

Reply:
While this is a noble thing we tried, and _should_ be working, we arrived to a libsolv issue/feature that was created for bad packagers, where an obsolete-ing package is considered as an upgrade candidate for the obsolete-d package (the tl;dr:). We can discuss this for later, but right now it works, and is user friendly enough IMO